### PR TITLE
Keepalive

### PR DESCRIPTION
**Pull request by mcasperson**

### CODE REVIEW

**Comment by mcasperson** [comment]
[Add keepalive option for Octo client](https://trello.com/c/eHs9rJWY/459-add-keepalive-option-for-octo-client)

File: source/Octopus.Cli/Commands/ApiCommand.cs [hunk 1]
```diff
@@ -127,6 +129,21 @@ public async Task Execute(string[] commandLineArguments)
...
+            
+            /*
+             * There may be a delay between the completion of a large file upload and when Octopus responds
+             * to finish the HTTP connection. This delay can be several minutes. During this time, no traffic is
+             * sent, and some networking infrastructure will close the connection. For example, Azure VMs will
+             * close idle connections after 4 minutes, and AWS VMs will close them after 350 seconds. The
+             * TCP keepalive option will ensure that the connection is not idle at the end of the file upload.
+             *
+             * This is the bug that explains why this doesn't work with .NET Core:
+             * https://github.com/dotnet/corefx/issues/26013
```
Comment (johnsimons) [inline source/Octopus.Cli/Commands/ApiCommand.cs]: @mcasperson given this has no effect in netcore should we add a compile directive to this and not even exposed this argument?

File: source/Octopus.Cli/Commands/ApiCommand.cs [hunk 2]
```diff
@@ -127,6 +129,21 @@ public async Task Execute(string[] commandLineArguments)
...
+             */
+            if (keepAlive > 0)
+            {
+                ServicePointManager.FindServicePoint(new Uri(ServerBaseUrl)).SetTcpKeepAlive(true, keepAlive, keepAlive);
```
Comment (johnsimons) [inline source/Octopus.Cli/Commands/ApiCommand.cs]: Is it safe to be setting both `keepAliveTime` and `keepAliveTimeInterval` to the same value?
Wouldn't you want `keepAliveTime` to be greater the `keepAliveTimeInterval`?

Comment (johnsimons) [reply source/Octopus.Cli/Commands/ApiCommand.cs]: Also, according to the API ( https://docs.microsoft.com/en-us/dotnet/api/system.net.servicepointmanager.settcpkeepalive) both `keepAliveTime` and `keepAliveTimeInterval` is set in milliseconds, but in https://github.com/OctopusDeploy/OctopusCLI/pull/11/files#diff-8f769fe53008cdd8986dbd9314ea7c95R74 we saying seconds?

**Comment by johnsimons** [comment]
I was just reading your comment @mcasperson 
> There may be a delay between the completion of a large file upload and when Octopus responds to finish the HTTP connection.

So are we masking a different issue by introducing this? Why is this delay occurring?

File: source/Octopus.Cli/Commands/ApiCommand.cs (hunk 2 above)
Comment (mcasperson) [reply source/Octopus.Cli/Commands/ApiCommand.cs]: `keepAliveTime` is the amount of time to elapse after the last packet was sent before keepalive is enabled, and then `keepAliveInterval` is the recurring time after that which packets are sent. Values in seconds were enough to keep the connection alive, and I didn't see any reason for the the values to be different. Is there a reason why` keepAliveTime` would be greater than `keepAliveTimeInterval`?

**Comment by mcasperson** [comment]
We are definitely going to look at the underlying cause, however given that enabling this setting fixed an issue for us, it could be used to fix issues for others with network topologies that we don't control. By making it an optional setting though we won't be masking anything by default.

File: source/Octopus.Cli/Commands/ApiCommand.cs (hunk 1 above)
Comment (mcasperson) [reply source/Octopus.Cli/Commands/ApiCommand.cs]: Thats a good idea, I've added a check to only expose it for the full .NET builds

**Comment by mcasperson** [comment]
> So this will only work for netfull, how about netcore?

It has not been exposed for the .NET Core libraries. This seems to be a limitation that does not have a work around.

File: source/Octopus.Cli/Commands/ApiCommand.cs (hunk 2 above)
Comment (johnsimons) [reply source/Octopus.Cli/Commands/ApiCommand.cs]: I just noticed you are  `* 1000`!
Regarding the 2 settings, honestly don't know, hence the reason to ask.

Comment (mcasperson) [reply source/Octopus.Cli/Commands/ApiCommand.cs]: For the purpose of keeping the connection alive while idea, I think both values achieve the same thing. Maybe there are more advanced networking scenarios that might benefit from different values, but I couldn't say what they would be.

**Pull request merged**